lib/ts/recipe/emailpassword/api/implementation.ts

```diff
@@ -682,47 +738,79 @@ export default function getAPIImplementation(): APIInterface {
-                    conflictingUsers.some((u) =>
-                        u.loginMethods.some((lm) => lm.recipeId === "emailpassword" && lm.hasSameEmailAs(email))
-                    )
-                ) {
+
+                if (!isSignUpAllowed) {
+                    const conflictingUsers = await AccountLinking.getInstance().recipeInterfaceImpl.listUsersByAccountInfo(
+                        {
+                            tenantId,
+                            accountInfo: {
+                                email,
+                            },
+                            doUnionOfAccountInfo: false,
+                            userContext,
+                        }
+                    );
+                    if (
+                        conflictingUsers.some((u) =>
+                            u.loginMethods.some((lm) => lm.recipeId === "emailpassword" && lm.hasSameEmailAs(email))
+                        )
+                    ) {
+                        return {
+                            status: "EMAIL_ALREADY_EXISTS_ERROR",
+                        };
+                    }
+
                     return {
-                        status: "EMAIL_ALREADY_EXISTS_ERROR",
+                        status: "SIGN_UP_NOT_ALLOWED",
+                        reason:
+                            "Cannot sign up due to security reasons. Please try logging in, use a different login method or contact support. (ERR_CODE_007)",
                     };
                 }
+            }
 
-                return {
-                    status: "SIGN_UP_NOT_ALLOWED",
-                    reason:
-                        "Cannot sign up due to security reasons. Please try logging in, use a different login method or contact support. (ERR_CODE_007)",
-                };
+            if (mfaInstance !== undefined) {
+                const mfaValidationRes = await mfaInstance.validateForMultifactorAuthBeforeFactorCompletion({
+                    req: options.req,
+                    res: options.res,
+                    tenantId,
+                    factorIdInProgress: "emailpassword",
+                    session,
+                    userLoggingIn: undefined,
+                    isAlreadySetup: false, // since this is a sign up
+                    signUpInfo: {
+                        email,
+                        isVerifiedFactor: false,
+                    },
+                    userContext,
+                });
+                if (mfaValidationRes.status !== "OK") {
+                    return mfaValidationRes;
+                }
             }
 
             // this function also does account linking
             let response = await options.recipeImplementation.signUp({
                 tenantId,
                 email,
                 password,
+                // we do not want to attempt accountlinking when there is an active session and MFA is turned on
+                shouldAttemptAccountLinkingIfAllowed: session === undefined,
```

Comment:
this is wrong. We also want to set it to false if true if mfa is not enabled, and the setting for session indicates that we can overwrite the session?

Reply:
Factor Flow is updated